server/src/main/java/org/elasticsearch/cluster/ClusterStateTaskExecutor.java

```diff
@@ -202,4 +202,75 @@ public Releasable dropHeadersContext() {
             return dropHeadersContextSupplier.get();
         }
     }
+
+    /**
+     * A basic implementation for batch executors that simply need to execute the tasks in the batch iteratively,
+     * producing a cluster state after each task. This allows executing the tasks in the batch as a
+     * series of executions, each taking an input cluster state and producing a new cluster state that serves as the
+     * input of the next task in the batch.
+     */
+    abstract class DefaultBatchExecutor<T extends ClusterStateTaskListener> implements ClusterStateTaskExecutor<T> {
+
+        /**
+         * Executes the provided task from the batch.
+         *
+         * @param taskContext The task to be executed.
+         * @param curState The cluster state on which the task should be executed.
+         * @return The resulting cluster state after executing this task.
+         */
+        public abstract ClusterState executeTask(TaskContext<T> taskContext, ClusterState curState);
+
+        /**
+         * Called once all tasks in the batch have finished execution. It should return a cluster state that reflects
+         * the execution of all the tasks.
+         *
+         * @param batchExecutionContext The tasks that were executed.
+         * @param initState The cluster state before executing the tasks.
+         * @param curState The cluster state resulting from the execution of all the tasks.
+         * @return The resulting cluster state after executing all the tasks.
+         */
+        public ClusterState afterBatchExecution(
+            BatchExecutionContext<T> batchExecutionContext,
+            ClusterState initState,
+            ClusterState curState
```

Comment:
Naming suggestion as above:

```suggestion
            ClusterState clusterState
```

(at the least we should call it `currentState` in full, no need to make it harder for non-English speakers for the sake of those 4 chars)